src/v/cluster/data_migration_backend.h

```diff
@@ -37,17 +41,130 @@ class backend {
+        explicit migration_reconciliation_state(state sought_state)
+          : sought_state(sought_state){};
+        state sought_state;
+        chunked_hash_map<model::topic_namespace, topic_reconciliation_state>
+          outstanding_topics;
+    };
+    using mrstates_t = absl::flat_hash_map<id, migration_reconciliation_state>;
 
-    ss::future<> reconcile_data_migration(id id);
-
-    chunked_hash_map<id, reconciliation_state> _states;
-    ss::gate _gate;
+    struct replica_work_state {
+        id migration_id;
+        state sought_state;
+        // shard may only be assigned if replica_status is can_run
+        std::optional<seastar::shard_id> shard;
+        migrated_replica_status status
+          = migrated_replica_status::waiting_for_rpc;
+        replica_work_state(id migration_id, state sought_state)
+          : migration_id(migration_id)
+          , sought_state(sought_state) {}
+    };
 
 private:
+    /* loop management */
+    ss::future<> loop_once();
+    ss::future<> work_once();
+    void wakeup();
+    [[nodiscard]] ss::future<std::optional<ssx::semaphore_units>> lock();
+
+    /* event handlers outside main loop */
+    ss::future<> handle_raft0_leadership_update();
+    ss::future<> handle_migration_update(id id);
+    void handle_shard_update(
+      const model::ntp& ntp, raft::group_id, std::optional<ss::shard_id> shard);
+
+    /* RPC and raft0 actions */
+    ss::future<> send_rpc(model::node_id node_id);
     ss::future<check_ntp_states_reply>
     check_ntp_states_locally(check_ntp_states_request&& req);
+    void to_advance(id migration_id, state sought_state);
+    void spawn_advances();
+
+    /* communication with workers */
+    void start_partition_work(
+      const model::ntp& ntp, const replica_work_state& rwstate);
+    void stop_partition_work(
+      const model::ntp& ntp, const replica_work_state& rwstate);
+    void
+    on_partition_work_completed(model::ntp&& ntp, id migration, state state);
+
+    /* deferred event handlers */
+    ss::future<> process_delta(cluster::topic_table_delta&& delta);
+
+    /* helpers */
+    void update_partition_shard(
+      const model::ntp& ntp,
+      replica_work_state& rwstate,
+      std::optional<ss::shard_id> new_shard);
+    void mark_rs_done_for_ntp(
+      migration_reconciliation_state& rs, const model::ntp& ntp);
+    void drop_mrstate(mrstates_t::const_iterator rs_it);
+    void clear_tstate_belongings(
+      const model::topic_namespace& nt,
+      const topic_reconciliation_state& tstate);
+
+    ss::future<> reconcile_migration(
+      migration_reconciliation_state& mrstate,
+      const migration_metadata& metadata);
+
+    ss::future<> reconcile_topic(
+      const model::topic_namespace& nt,
+      topic_reconciliation_state& tstate,
+      id migration,
+      state sought_state,
+      bool schedule_local_work);
+
+    std::optional<std::reference_wrapper<replica_work_state>>
+    get_rwstate(const model::ntp& ntp);
```

Comment:
nit: the names of methods and fields of the backend class are very short, consider making them a bit more verbose to better reflect their meaning. The code will be easier to read this way